interchain-token-service/src/contract/execute.rs

```diff
@@ -0,0 +1,329 @@
+}
+
+/// Executes an incoming ITS message.
+///
+/// This function handles the execution of ITS (Interchain Token Service) messages received from
+/// its sources. It verifies the source address, decodes the message, applies balance tracking,
+/// and forwards the message to the destination chain.
+pub fn execute_message(
+    deps: DepsMut,
+    cc_id: CrossChainId,
+    source_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    ensure_its_source_address(deps.storage, &cc_id.source_chain, &source_address)?;
+
+    match ItsHubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)? {
+        ItsHubMessage::SendToHub {
+            destination_chain,
+            message,
+        } => {
+            let destination_chain = to_chain_name(&destination_chain);
+            let destination_address = load_its_address(deps.storage, &destination_chain)
+                .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
+
+            let destination_payload = ItsHubMessage::ReceiveFromHub {
+                source_chain: cc_id.source_chain.clone(),
+                message: message.clone(),
+            }
+            .abi_encode();
+
+            Ok(send_to_destination(
+                deps.storage,
+                deps.querier,
+                destination_chain.clone(),
+                destination_address,
+                destination_payload,
+            )?
+            .add_event(
+                ItsContractEvent::ItsMessageReceived {
+                    cc_id,
+                    destination_chain,
+                    message,
+                }
+                .into(),
+            ))
+        }
+        _ => bail!(Error::InvalidPayload),
+    }
+}
+
+fn to_chain_name(chain: &ChainNameRaw) -> ChainName {
+    ChainName::try_from(chain.as_ref()).expect("invalid chain name")
+}
+
+fn ensure_its_source_address(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    source_address: &Address,
+) -> Result<(), Error> {
+    let source_chain = to_chain_name(source_chain);
+    let its_source_address = load_its_address(storage, &source_chain)
+        .change_context_lazy(|| Error::UnknownChain(source_chain.clone()))?;
+
+    ensure!(
+        source_address == &its_source_address,
+        Error::UnknownItsAddress(source_address.clone())
+    );
+
+    Ok(())
+}
+
+fn send_to_destination(
+    storage: &dyn Storage,
+    querier: QuerierWrapper,
+    destination_chain: ChainName,
+    destination_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let config = load_config(storage).change_context(Error::ConfigAccess)?;
+
+    let gateway: axelarnet_gateway::Client =
+        client::Client::new(querier, config.axelarnet_gateway).into();
+
+    let call_contract_msg = gateway.call_contract(destination_chain, destination_address, payload);
+
+    Ok(Response::new().add_message(call_contract_msg))
+}
+
+pub fn set_its_address(
```

Comment:
I think it makes more sense to call all these references to `its_address` `its_contract_address` instead

Reply:
`its_address` being a contract feels implied, other potential names: `external_its_address`, `its_edge_address`, `remote_its_address`.